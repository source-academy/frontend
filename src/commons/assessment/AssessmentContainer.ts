import { connect, MapDispatchToProps, MapStateToProps } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';

import {
  acknowledgeNotifications,
  fetchAssessmentOverviews,
  submitAssessment
} from '../application/actions/SessionActions';
import { OverallState, Role } from '../application/ApplicationTypes';
import Assessment, { DispatchProps, OwnProps, StateProps } from './Assessment';

const mapStateToProps: MapStateToProps<StateProps, OwnProps, OverallState> = (state, props) => ({
  assessmentOverviews: state.session.assessmentOverviews,
  isStudent: state.session.role ? state.session.role === Role.Student : true,
  courseId: state.session.courseId
});
const mapDispatchToProps: MapDispatchToProps<DispatchProps, {}> = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      handleAcknowledgeNotifications: acknowledgeNotifications,
      handleAssessmentOverviewFetch: fetchAssessmentOverviews,
      handleSubmitAssessment: submitAssessment
    },
    dispatch
  );

const AssessmentContainer = connect(mapStateToProps, mapDispatchToProps)(Assessment);

export default AssessmentContainer;
