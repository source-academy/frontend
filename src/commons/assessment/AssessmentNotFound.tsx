import { Classes, NonIdealState } from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import classNames from 'classnames';

const AssessmentNotFound: React.FC = () => (
  <div className={classNames('NoPage', Classes.DARK)}>
    <NonIdealState
      icon={IconNames.ERROR}
      title="Assessment Not Found"
      description="Please check the top right dropdown menu to see if you are in the right course."
    />
  </div>
);

export default AssessmentNotFound;
