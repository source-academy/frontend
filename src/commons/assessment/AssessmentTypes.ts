import { Chapter, SourceError, Variant } from 'js-slang/dist/types';

import { ExternalLibrary, ExternalLibraryName } from '../application/types/ExternalTypes';

export const FETCH_ASSESSMENT_OVERVIEWS = 'FETCH_ASSESSMENT_OVERVIEWS';
export const SUBMIT_ASSESSMENT = 'SUBMIT_ASSESSMENT';

export enum AssessmentStatuses {
  attempting = 'attempting',
  attempted = 'attempted',
  not_attempted = 'not_attempted',
  submitted = 'submitted'
}
export type AssessmentStatus = keyof typeof AssessmentStatuses;

export type AssessmentWorkspaceParams = {
  assessmentId?: string;
  questionId?: string;
};

export enum GradingStatuses {
  excluded = 'excluded',
  graded = 'graded',
  grading = 'grading',
  none = 'none'
}
export type GradingStatus = keyof typeof GradingStatuses;

export type AssessmentType = string;

export enum TestcaseTypes {
  // These are rendered in full by the Mission Autograder
  public = 'public',
  // These are rendered with a placeholder by the Autograder
  opaque = 'opaque',
  // These should only exist in the grading workspace for submissions
  secret = 'secret'
}
export type TestcaseType = keyof typeof TestcaseTypes;

export enum QuestionTypes {
  programming = 'programming',
  mcq = 'mcq',
  voting = 'voting'
}
export type QuestionType = keyof typeof QuestionTypes;

/*
W* Used to display information regarding an assessment in the UI.
*
 * @property closeAt an ISO 8601 compliant date string specifiying when
 *   the assessment closes
 * @property openAt an ISO 8601 compliant date string specifiying when
 *   the assessment opens
 */
export type AssessmentOverview = {
  type: AssessmentType;
  isManuallyGraded: boolean;
  closeAt: string;
  coverImage: string;
  fileName?: string; // For mission control
  gradingStatus: GradingStatus;
  id: number;
  isPublished?: boolean;
  maxXp: number;
  number?: string; // For mission control
  openAt: string;
  private?: boolean;
  reading?: string; // For mission control
  shortSummary: string;
  status: AssessmentStatus;
  story: string | null;
  title: string;
  xp: number;
};

/*
 * Used when an assessment is being actively attempted/graded.
 */
export type Assessment = {
  type: AssessmentType;
  globalDeployment?: Library; // For mission control
  graderDeployment?: Library; // For mission control
  id: number;
  longSummary: string;
  missionPDF: string;
  title: string;
  questions: Question[];
};

export type AssessmentConfiguration = {
  assessmentConfigId: number;
  type: AssessmentType;
  isManuallyGraded: boolean;
  displayInDashboard: boolean;
  hoursBeforeEarlyXpDecay: number;
  earlySubmissionXp: number;
};

export interface IProgrammingQuestion extends BaseQuestion {
  answer: string | null;
  autogradingResults: AutogradingResult[];
  graderTemplate?: string;
  prepend: string;
  postpend: string;
  solutionTemplate: string;
  testcases: Testcase[];
  testcasesPrivate?: Testcase[]; // For mission control
  type: 'programming';
}

export interface IMCQQuestion extends BaseQuestion {
  answer: number | null;
  choices: MCQChoice[];
  solution?: number;
  type: 'mcq';
}

export interface IContestVotingQuestion extends BaseQuestion {
  answer: ContestEntry[];
  prepend: string;
  postpend: string;
  contestEntries: ContestEntry[];
  contestLeaderboard: ContestEntry[];
  type: 'voting';
}

export type BaseQuestion = {
  answer: string | number | ContestEntry[] | null;
  comments?: string;
  content: string;
  editorValue?: string;
  gradedAt?: string;
  grader?: {
    name: string;
    id: number;
  };
  graderLibrary?: Library; // For mission control
  id: number;
  library: Library;
  maxXp: number;
  type: QuestionType;
  xp: number;
  blocking?: boolean; // Determines whether the learner can progress to the next question without passing local testcases
  // TODO: The blocking field is made optional now as the Question type is being shared with GitHub Assessments, which has not implemented
  // the question-level blocking feature. Is to be made compulsory after this is implemented in GitHub Assessments
};

export type Question = IProgrammingQuestion | IMCQQuestion | IContestVotingQuestion;

export type Library = {
  chapter: Chapter;
  variant?: Variant;
  execTimeMs?: number;
  external: ExternalLibrary;
  globals: Array<{
    0: string;
    1: any;
    2?: string; // For mission control
  }>;
  moduleParams?: any;
};

export type Testcase = {
  answer: string; // the correct answer to the testcase
  errors?: SourceError[]; // errors raised by interpreter during execution
  program: string; // the program to be appended to the student's code
  result?: any; // the result from the execution of the testcase
  score: number;
  type: TestcaseType;
};

export type ContestEntry = {
  submission_id: number;
  answer: ContestEntryCodeAnswer; //contest entry program to be input into editor
  score?: number;
  final_score?: number;
  student_name?: string;
};

export type ContestEntryCodeAnswer = {
  code: string;
};

export type MCQChoice = {
  content: string;
  hint: string | null;
};

export type AutogradingResult = {
  resultType: string;
  expected?: string; // the correct answer for the testcase
  actual?: string; // the received answer from the student's code
  errors?: AutogradingError[];
};

export type AutogradingError = {
  errorType: string;
  line?: number;
  location?: string;
  errorLine?: string;
  errorExplanation?: string;
  errorMessage?: string;
};

export const emptyLibrary = (): Library => {
  return {
    chapter: -1,
    external: {
      name: 'NONE' as ExternalLibraryName,
      symbols: []
    },
    globals: []
  };
};

export const normalLibrary = (): Library => {
  return {
    chapter: Chapter.SOURCE_1,
    external: {
      name: 'NONE' as ExternalLibraryName,
      symbols: []
    },
    globals: []
  };
};

export const overviewTemplate = (): AssessmentOverview => {
  return {
    type: 'Missions',
    isManuallyGraded: true,
    closeAt: '2100-12-01T00:00+08',
    coverImage: 'https://fakeimg.pl/300/',
    id: -1,
    maxXp: 0,
    openAt: '2000-01-01T00:00+08',
    title: 'Insert title here',
    reading: '',
    shortSummary: 'Insert short summary here',
    status: AssessmentStatuses.not_attempted,
    story: 'mission',
    xp: 0,
    gradingStatus: 'none'
  };
};

export const programmingTemplate = (): IProgrammingQuestion => {
  return {
    autogradingResults: [],
    answer: '// [Marking Scheme]\n// 1 mark for correct answer',
    content: 'Enter content here',
    id: 0,
    library: emptyLibrary(),
    graderLibrary: emptyLibrary(),
    prepend: '',
    solutionTemplate: '//This is a mock solution template',
    postpend: '',
    testcases: [],
    testcasesPrivate: [],
    type: 'programming',
    xp: 0,
    maxXp: 0,
    blocking: false
  };
};

export const testcaseTemplate = (): Testcase => {
  return {
    type: TestcaseTypes.public,
    answer: '',
    score: 0,
    program: ''
  };
};

export const mcqTemplate = (): IMCQQuestion => {
  return {
    answer: 3,
    content: 'This is a mock MCQ question',
    choices: [
      {
        content: 'A',
        hint: null
      },
      {
        content: 'B',
        hint: null
      },
      {
        content: 'C',
        hint: null
      },
      {
        content: 'D',
        hint: null
      }
    ],
    id: 2,
    library: emptyLibrary(),
    graderLibrary: emptyLibrary(),
    type: 'mcq',
    solution: 0,
    xp: 0,
    maxXp: 0,
    blocking: false
  };
};

export const assessmentTemplate = (): Assessment => {
  return {
    type: 'Missions',
    globalDeployment: normalLibrary(),
    graderDeployment: emptyLibrary(),
    id: -1,
    longSummary: 'Insert mission briefing here',
    missionPDF: 'www.google.com',
    questions: [programmingTemplate()],
    title: 'Insert title here'
  };
};
