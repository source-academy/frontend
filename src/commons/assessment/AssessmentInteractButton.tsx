import { Button } from '@blueprintjs/core';
import { IconName, IconNames } from '@blueprintjs/icons';
import { useDispatch } from 'react-redux';
import { NavLink } from 'react-router';

import SessionActions from '../application/actions/SessionActions';
import { filterNotificationsByAssessment } from '../notificationBadge/NotificationBadgeHelper';
import Constants from '../utils/Constants';
import { useTypedSelector } from '../utils/Hooks';
import { assessmentTypeLink } from '../utils/ParamParseHelper';
import { AssessmentOverview, AssessmentStatuses } from './AssessmentTypes';

type Props = {
  overview: AssessmentOverview;
};

const AssessmentInteractButton: React.FC<Props> = ({ overview }) => {
  const courseId = useTypedSelector(state => state.session.courseId);
  const dispatch = useDispatch();
  const { icon, label, optionalLabel } = createButtonConfiguration(overview.status);

  return (
    <NavLink
      to={`/courses/${courseId}/${assessmentTypeLink(overview.type)}/${overview.id.toString()}/${
        Constants.defaultQuestionId
      }`}
    >
      <Button
        icon={icon}
        variant="minimal"
        onClick={() =>
          dispatch(
            SessionActions.acknowledgeNotifications(filterNotificationsByAssessment(overview.id))
          )
        }
      >
        <span data-testid="Assessment-Attempt-Button">{label}</span>
        {optionalLabel && <span className="custom-hidden-xxxs">{optionalLabel}</span>}
      </Button>
    </NavLink>
  );
};

type ButtonConfiguration = {
  icon: IconName;
  label: string;
  optionalLabel?: string;
};

const createButtonConfiguration = (
  overviewStatus: AssessmentOverview['status']
): ButtonConfiguration => {
  let icon: IconName;
  let label: string;
  let optionalLabel: string | undefined;

  switch (overviewStatus) {
    case AssessmentStatuses.not_attempted:
      icon = IconNames.PLAY;
      label = 'Attempt';
      break;
    case AssessmentStatuses.attempting:
      icon = IconNames.PLAY;
      label = 'Continue';
      optionalLabel = ' Attempt';
      break;
    case AssessmentStatuses.attempted:
      icon = IconNames.EDIT;
      label = 'Review';
      optionalLabel = ' Attempt';
      break;
    case AssessmentStatuses.submitted:
      icon = IconNames.EYE_OPEN;
      label = 'Review';
      optionalLabel = ' Submission';
      break;
    default:
      // If we reach this case, backend data did not fit IAssessmentOverview
      icon = IconNames.PLAY;
      label = 'Review';
      break;
  }

  return { icon, label, optionalLabel };
};

export default AssessmentInteractButton;
