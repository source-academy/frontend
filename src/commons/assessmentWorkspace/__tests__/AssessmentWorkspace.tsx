import { shallow } from 'enzyme';

import { ContestEntry, Library } from '../../assessment/AssessmentTypes';
import { EditorProps } from '../../editor/Editor';
import { Position } from '../../editor/EditorTypes';
import { mockAssessments } from '../../mocks/AssessmentMocks';
import AssessmentWorkspace, { AssessmentWorkspaceProps } from '../AssessmentWorkspace';
const MockEditor = (props: EditorProps) => <div id="mock-editor">{props.editorValue}</div>;
// mock editor for testing update
jest.mock('../../editor/Editor', () => (props: EditorProps) => (
  <MockEditor {...props}></MockEditor>
));

const mockedHandleEditorValueChange = jest.fn();

const defaultProps: AssessmentWorkspaceProps = {
  assessmentId: 0,
  autogradingResults: [],
  notAttempted: true,
  canSave: true,
  assessmentConfiguration: {
    assessmentConfigId: 1,
    type: 'Missions',
    isManuallyGraded: true,
    displayInDashboard: true,
    hoursBeforeEarlyXpDecay: 48,
    earlySubmissionXp: 200
  },
  editorPrepend: '',
  editorValue: null,
  editorPostpend: '',
  editorTestcases: [],
  editorWidth: '50%',
  breakpoints: [],
  highlightedLines: [],
  hasUnsavedChanges: false,
  handleAssessmentFetch: (assessmentId: number) => {},
  handleBrowseHistoryDown: () => {},
  handleBrowseHistoryUp: () => {},
  handleClearContext: (library: Library, shouldInitLibrary: boolean) => {},
  handleDeclarationNavigate: (cursorPosition: Position) => {},
  handleEditorEval: () => {},
  handleEditorValueChange: mockedHandleEditorValueChange,
  handleEditorHeightChange: (height: number) => {},
  handleEditorWidthChange: (widthChange: number) => {},
  handleEditorUpdateBreakpoints: (breakpoints: string[]) => {},
  handleInterruptEval: () => {},
  handleReplEval: () => {},
  handleReplOutputClear: () => {},
  handleReplValueChange: (newValue: string) => {},
  handleSendReplInputToOutput: (code: string) => {},
  handleResetWorkspace: () => {},
  handleSave: (id: number, answer: number | string | ContestEntry[]) => {},
  handleSideContentHeightChange: (heightChange: number) => {},
  handleTestcaseEval: (testcaseId: number) => {},
  handleRunAllTestcases: () => {},
  handleUpdateHasUnsavedChanges: (hasUnsavedChanges: boolean) => {},
  handleUpdateCurrentAssessmentId: (a: number, q: number) => {},
  handleDebuggerPause: () => {},
  handleDebuggerResume: () => {},
  handleDebuggerReset: () => {},
  handlePromptAutocomplete: (row: number, col: number, callback: any) => {},
  isRunning: false,
  isDebugging: false,
  enableDebugging: false,
  output: [],
  questionId: 0,
  replValue: ''
};

const mockUndefinedAssessmentWorkspaceProps: AssessmentWorkspaceProps = {
  ...defaultProps
};

const mockProgrammingAssessmentWorkspaceProps: AssessmentWorkspaceProps = {
  ...defaultProps,
  assessment: mockAssessments[0],
  assessmentId: 0,
  questionId: 0
};

const mockClosedProgrammingAssessmentWorkspaceProps: AssessmentWorkspaceProps = {
  ...mockProgrammingAssessmentWorkspaceProps,
  canSave: false
};

const mockGradedProgrammingAssessmentWorkspaceProps: AssessmentWorkspaceProps = {
  ...defaultProps,
  assessment: mockAssessments[3],
  assessmentId: 4,
  questionId: 0
};

const mockMcqAssessmentWorkspaceProps: AssessmentWorkspaceProps = {
  ...defaultProps,
  assessment: mockAssessments[0],
  assessmentId: 0,
  questionId: 2
};

// set questionId to index 0 since contest voting only has 1 question
const mockContestVotingAssessmentWorkspaceProps: AssessmentWorkspaceProps = {
  ...defaultProps,
  assessment: mockAssessments[6],
  assessmentId: 7,
  questionId: 0
};

test('AssessmentWorkspace page "loading" content renders correctly', () => {
  const app = <AssessmentWorkspace {...mockUndefinedAssessmentWorkspaceProps} />;
  const tree = shallow(app);
  expect(tree.debug()).toMatchSnapshot();
});

test('AssessmentWorkspace page with programming question renders correctly', () => {
  const app = <AssessmentWorkspace {...mockProgrammingAssessmentWorkspaceProps} />;
  const tree = shallow(app);
  expect(tree.debug()).toMatchSnapshot();
});

test('AssessmentWorkspace page with overdue assessment renders correctly', () => {
  const app = <AssessmentWorkspace {...mockClosedProgrammingAssessmentWorkspaceProps} />;
  const tree = shallow(app);
  expect(tree.debug()).toMatchSnapshot();
});

test('AssessmentWorkspace page with MCQ question renders correctly', () => {
  const app = <AssessmentWorkspace {...mockMcqAssessmentWorkspaceProps} />;
  const tree = shallow(app);
  expect(tree.debug()).toMatchSnapshot();
});

/*  ===== Tester comments =====
    Issue:
      https://stackoverflow.com/questions/42813342/react-createelement-type-is-invalid-expected-a-string
    Description:
      Mounting the AssessmentWorkspace deeply (rendering all recursive subcomponents) in Enzyme with mount
      results in the test failing with the error
          Warning: React.createElement: type is invalid -- expected a string (for built-in components) or
          a class/function (for composite components) but got: undefined. You likely forgot to export your
          component from the file it's defined in, or you might have mixed up default and named imports.

          Check the render method of `Workspace`.

          The above error occurred in the <div> component:
            in div (created by Workspace)
            in div (created by Workspace)
            in Workspace (created by AssessmentWorkspace)
            in div (created by AssessmentWorkspace)
            in AssessmentWorkspace (created by WrapperComponent)
            in WrapperComponent

      whereas mounting it one-level deep in Enzyme using shallow throws no errors
    Fix:
      Stack trace suggests one of the React subcomponents of AssessmentWorkspace works in production
      but is not set up correctly - requires re-examination of every single React component and
      sub-component used in AssessmentWorkspace

      Current workaround is to mount AssessmentWorkspace shallowly since the behaviour is correct
      during user testing
*/

test('AssessmentWorkspace page with ContestVoting question renders correctly', () => {
  const app = <AssessmentWorkspace {...mockContestVotingAssessmentWorkspaceProps} />;
  const tree = shallow(app);
  expect(tree.debug()).toMatchSnapshot();
});

test('AssessmentWorkspace renders Grading tab correctly if the question has been graded', () => {
  const app = <AssessmentWorkspace {...mockGradedProgrammingAssessmentWorkspaceProps} />;
  const tree = shallow(app);
  expect(tree.debug()).toMatchSnapshot();
  // Uncomment when fixed
  // expect(tree.find('.grading-icon').hostNodes()).toHaveLength(1);
});
