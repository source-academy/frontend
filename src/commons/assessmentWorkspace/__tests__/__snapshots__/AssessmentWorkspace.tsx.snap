// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssessmentWorkspace page "loading" content renders correctly 1`] = `"<Blueprint4.NonIdealState className=\\"WorkspaceParent bp4-dark\\" description=\\"Getting mission ready...\\" icon={{...}} iconSize={48} layout=\\"vertical\\" />"`;

exports[`AssessmentWorkspace page with ContestVoting question renders correctly 1`] = `
"<div className=\\"WorkspaceParent bp4-dark\\">
  <Blueprint4.Dialog className=\\"assessment-briefing\\" isOpen={false} canOutsideClickClose={true}>
    <Blueprint4.Card elevation={0} interactive={false}>
      <Memo(Markdown) content=\\"Vote for your favourite contest entries here!\\" />
      <Blueprint4.Button className=\\"assessment-briefing-button\\" onClick={[Function: onClick]} text=\\"Continue\\" />
    </Blueprint4.Card>
  </Blueprint4.Dialog>
  <Blueprint4.Dialog className=\\"assessment-reset\\" icon=\\"error\\" isCloseButtonShown={true} isOpen={false} onClose={[Function: closeOverlay]} title=\\"Confirmation: Reset editor?\\" canOutsideClickClose={true}>
    <div className=\\"bp4-dialog-body\\">
      <Memo(Markdown) content=\\"Are you sure you want to reset the template?\\" />
      <Memo(Markdown) content=\\"*Note this will not affect the saved copy of your program, unless you save over it.*\\" />
    </div>
    <div className=\\"bp4-dialog-footer\\">
      <Blueprint4.ButtonGroup>
        <Blueprint4.Button disabled={false} fill={false} intent=\\"none\\" minimal={false} className=\\"\\" type={[undefined]} onClick={[Function: closeOverlay]}>
          Cancel
        </Blueprint4.Button>
        <Blueprint4.Button disabled={false} fill={false} intent=\\"danger\\" minimal={false} className=\\"\\" type={[undefined]} onClick={[Function (anonymous)]}>
          Confirm
        </Blueprint4.Button>
      </Blueprint4.ButtonGroup>
    </div>
  </Blueprint4.Dialog>
  <Workspace controlBarProps={{...}} editorProps={{...}} editorHeight={[undefined]} editorWidth=\\"50%\\" handleEditorHeightChange={[Function: handleEditorHeightChange]} handleEditorWidthChange={[Function: handleEditorWidthChange]} handleSideContentHeightChange={[Function: handleSideContentHeightChange]} hasUnsavedChanges={false} mcqProps={{...}} sideContentHeight={[undefined]} sideContentProps={{...}} replProps={{...}} />
</div>"
`;

exports[`AssessmentWorkspace page with MCQ question renders correctly 1`] = `
"<div className=\\"WorkspaceParent bp4-dark\\">
  <Blueprint4.Dialog className=\\"assessment-briefing\\" isOpen={false} canOutsideClickClose={true}>
    <Blueprint4.Card elevation={0} interactive={false}>
      <Memo(Markdown) content=\\"This is the mission briefing. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Maecenas viverra, sem scelerisque ultricies ullamcorper, sem nibh sollicitudin enim, at ultricies sem orci eget odio. Pellentesque varius et mauris quis vestibulum. Etiam in egestas dolor. Nunc consectetur, sapien sodales accumsan convallis, lectus mi tempus ipsum, vel ornare metus turpis sed justo. Vivamus at tellus sed ex convallis commodo at in lectus. Pellentesque pharetra pulvinar sapien pellentesque facilisis. Curabitur efficitur malesuada urna sed aliquam. Quisque massa metus, aliquam in sagittis non, cursus in sem. Morbi vel nunc at nunc pharetra lobortis. Aliquam feugiat ultricies ipsum vel sollicitudin. Vivamus nulla massa, hendrerit sit amet nibh quis, porttitor convallis nisi. \\" />
      <Blueprint4.Button className=\\"assessment-briefing-button\\" onClick={[Function: onClick]} text=\\"Continue\\" />
    </Blueprint4.Card>
  </Blueprint4.Dialog>
  <Blueprint4.Dialog className=\\"assessment-reset\\" icon=\\"error\\" isCloseButtonShown={true} isOpen={false} onClose={[Function: closeOverlay]} title=\\"Confirmation: Reset editor?\\" canOutsideClickClose={true}>
    <div className=\\"bp4-dialog-body\\">
      <Memo(Markdown) content=\\"Are you sure you want to reset the template?\\" />
      <Memo(Markdown) content=\\"*Note this will not affect the saved copy of your program, unless you save over it.*\\" />
    </div>
    <div className=\\"bp4-dialog-footer\\">
      <Blueprint4.ButtonGroup>
        <Blueprint4.Button disabled={false} fill={false} intent=\\"none\\" minimal={false} className=\\"\\" type={[undefined]} onClick={[Function: closeOverlay]}>
          Cancel
        </Blueprint4.Button>
        <Blueprint4.Button disabled={false} fill={false} intent=\\"danger\\" minimal={false} className=\\"\\" type={[undefined]} onClick={[Function (anonymous)]}>
          Confirm
        </Blueprint4.Button>
      </Blueprint4.ButtonGroup>
    </div>
  </Blueprint4.Dialog>
  <Workspace controlBarProps={{...}} editorProps={[undefined]} editorHeight={[undefined]} editorWidth=\\"50%\\" handleEditorHeightChange={[Function: handleEditorHeightChange]} handleEditorWidthChange={[Function: handleEditorWidthChange]} handleSideContentHeightChange={[Function: handleSideContentHeightChange]} hasUnsavedChanges={false} mcqProps={{...}} sideContentHeight={[undefined]} sideContentProps={{...}} replProps={{...}} />
</div>"
`;

exports[`AssessmentWorkspace page with overdue assessment renders correctly 1`] = `
"<div className=\\"WorkspaceParent bp4-dark\\">
  <Blueprint4.Dialog className=\\"assessment-briefing\\" isOpen={false} canOutsideClickClose={true}>
    <Blueprint4.Card elevation={0} interactive={false}>
      <Memo(Markdown) content=\\"This is the mission briefing. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Maecenas viverra, sem scelerisque ultricies ullamcorper, sem nibh sollicitudin enim, at ultricies sem orci eget odio. Pellentesque varius et mauris quis vestibulum. Etiam in egestas dolor. Nunc consectetur, sapien sodales accumsan convallis, lectus mi tempus ipsum, vel ornare metus turpis sed justo. Vivamus at tellus sed ex convallis commodo at in lectus. Pellentesque pharetra pulvinar sapien pellentesque facilisis. Curabitur efficitur malesuada urna sed aliquam. Quisque massa metus, aliquam in sagittis non, cursus in sem. Morbi vel nunc at nunc pharetra lobortis. Aliquam feugiat ultricies ipsum vel sollicitudin. Vivamus nulla massa, hendrerit sit amet nibh quis, porttitor convallis nisi. \\" />
      <Blueprint4.Button className=\\"assessment-briefing-button\\" onClick={[Function: onClick]} text=\\"Continue\\" />
    </Blueprint4.Card>
  </Blueprint4.Dialog>
  <Blueprint4.Dialog className=\\"assessment-reset\\" icon=\\"error\\" isCloseButtonShown={true} isOpen={false} onClose={[Function: closeOverlay]} title=\\"Confirmation: Reset editor?\\" canOutsideClickClose={true}>
    <div className=\\"bp4-dialog-body\\">
      <Memo(Markdown) content=\\"Are you sure you want to reset the template?\\" />
      <Memo(Markdown) content=\\"*Note this will not affect the saved copy of your program, unless you save over it.*\\" />
    </div>
    <div className=\\"bp4-dialog-footer\\">
      <Blueprint4.ButtonGroup>
        <Blueprint4.Button disabled={false} fill={false} intent=\\"none\\" minimal={false} className=\\"\\" type={[undefined]} onClick={[Function: closeOverlay]}>
          Cancel
        </Blueprint4.Button>
        <Blueprint4.Button disabled={false} fill={false} intent=\\"danger\\" minimal={false} className=\\"\\" type={[undefined]} onClick={[Function (anonymous)]}>
          Confirm
        </Blueprint4.Button>
      </Blueprint4.ButtonGroup>
    </div>
  </Blueprint4.Dialog>
  <Workspace controlBarProps={{...}} editorProps={{...}} editorHeight={[undefined]} editorWidth=\\"50%\\" handleEditorHeightChange={[Function: handleEditorHeightChange]} handleEditorWidthChange={[Function: handleEditorWidthChange]} handleSideContentHeightChange={[Function: handleSideContentHeightChange]} hasUnsavedChanges={false} mcqProps={{...}} sideContentHeight={[undefined]} sideContentProps={{...}} replProps={{...}} />
</div>"
`;

exports[`AssessmentWorkspace page with programming question renders correctly 1`] = `
"<div className=\\"WorkspaceParent bp4-dark\\">
  <Blueprint4.Dialog className=\\"assessment-briefing\\" isOpen={false} canOutsideClickClose={true}>
    <Blueprint4.Card elevation={0} interactive={false}>
      <Memo(Markdown) content=\\"This is the mission briefing. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Maecenas viverra, sem scelerisque ultricies ullamcorper, sem nibh sollicitudin enim, at ultricies sem orci eget odio. Pellentesque varius et mauris quis vestibulum. Etiam in egestas dolor. Nunc consectetur, sapien sodales accumsan convallis, lectus mi tempus ipsum, vel ornare metus turpis sed justo. Vivamus at tellus sed ex convallis commodo at in lectus. Pellentesque pharetra pulvinar sapien pellentesque facilisis. Curabitur efficitur malesuada urna sed aliquam. Quisque massa metus, aliquam in sagittis non, cursus in sem. Morbi vel nunc at nunc pharetra lobortis. Aliquam feugiat ultricies ipsum vel sollicitudin. Vivamus nulla massa, hendrerit sit amet nibh quis, porttitor convallis nisi. \\" />
      <Blueprint4.Button className=\\"assessment-briefing-button\\" onClick={[Function: onClick]} text=\\"Continue\\" />
    </Blueprint4.Card>
  </Blueprint4.Dialog>
  <Blueprint4.Dialog className=\\"assessment-reset\\" icon=\\"error\\" isCloseButtonShown={true} isOpen={false} onClose={[Function: closeOverlay]} title=\\"Confirmation: Reset editor?\\" canOutsideClickClose={true}>
    <div className=\\"bp4-dialog-body\\">
      <Memo(Markdown) content=\\"Are you sure you want to reset the template?\\" />
      <Memo(Markdown) content=\\"*Note this will not affect the saved copy of your program, unless you save over it.*\\" />
    </div>
    <div className=\\"bp4-dialog-footer\\">
      <Blueprint4.ButtonGroup>
        <Blueprint4.Button disabled={false} fill={false} intent=\\"none\\" minimal={false} className=\\"\\" type={[undefined]} onClick={[Function: closeOverlay]}>
          Cancel
        </Blueprint4.Button>
        <Blueprint4.Button disabled={false} fill={false} intent=\\"danger\\" minimal={false} className=\\"\\" type={[undefined]} onClick={[Function (anonymous)]}>
          Confirm
        </Blueprint4.Button>
      </Blueprint4.ButtonGroup>
    </div>
  </Blueprint4.Dialog>
  <Workspace controlBarProps={{...}} editorProps={{...}} editorHeight={[undefined]} editorWidth=\\"50%\\" handleEditorHeightChange={[Function: handleEditorHeightChange]} handleEditorWidthChange={[Function: handleEditorWidthChange]} handleSideContentHeightChange={[Function: handleSideContentHeightChange]} hasUnsavedChanges={false} mcqProps={{...}} sideContentHeight={[undefined]} sideContentProps={{...}} replProps={{...}} />
</div>"
`;

exports[`AssessmentWorkspace renders Grading tab correctly if the question has been graded 1`] = `
"<div className=\\"WorkspaceParent bp4-dark\\">
  <Blueprint4.Dialog className=\\"assessment-briefing\\" isOpen={false} canOutsideClickClose={true}>
    <Blueprint4.Card elevation={0} interactive={false}>
      <Memo(Markdown) content=\\"This is the closed mission briefing. The save button should not be there. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Maecenas viverra, sem scelerisque ultricies ullamcorper, sem nibh sollicitudin enim, at ultricies sem orci eget odio. Pellentesque varius et mauris quis vestibulum. Etiam in egestas dolor. Nunc consectetur, sapien sodales accumsan convallis, lectus mi tempus ipsum, vel ornare metus turpis sed justo. Vivamus at tellus sed ex convallis commodo at in lectus. Pellentesque pharetra pulvinar sapien pellentesque facilisis. Curabitur efficitur malesuada urna sed aliquam. Quisque massa metus, aliquam in sagittis non, cursus in sem. Morbi vel nunc at nunc pharetra lobortis. Aliquam feugiat ultricies ipsum vel sollicitudin. Vivamus nulla massa, hendrerit sit amet nibh quis, porttitor convallis nisi. \\" />
      <Blueprint4.Button className=\\"assessment-briefing-button\\" onClick={[Function: onClick]} text=\\"Continue\\" />
    </Blueprint4.Card>
  </Blueprint4.Dialog>
  <Blueprint4.Dialog className=\\"assessment-reset\\" icon=\\"error\\" isCloseButtonShown={true} isOpen={false} onClose={[Function: closeOverlay]} title=\\"Confirmation: Reset editor?\\" canOutsideClickClose={true}>
    <div className=\\"bp4-dialog-body\\">
      <Memo(Markdown) content=\\"Are you sure you want to reset the template?\\" />
      <Memo(Markdown) content=\\"*Note this will not affect the saved copy of your program, unless you save over it.*\\" />
    </div>
    <div className=\\"bp4-dialog-footer\\">
      <Blueprint4.ButtonGroup>
        <Blueprint4.Button disabled={false} fill={false} intent=\\"none\\" minimal={false} className=\\"\\" type={[undefined]} onClick={[Function: closeOverlay]}>
          Cancel
        </Blueprint4.Button>
        <Blueprint4.Button disabled={false} fill={false} intent=\\"danger\\" minimal={false} className=\\"\\" type={[undefined]} onClick={[Function (anonymous)]}>
          Confirm
        </Blueprint4.Button>
      </Blueprint4.ButtonGroup>
    </div>
  </Blueprint4.Dialog>
  <Workspace controlBarProps={{...}} editorProps={{...}} editorHeight={[undefined]} editorWidth=\\"50%\\" handleEditorHeightChange={[Function: handleEditorHeightChange]} handleEditorWidthChange={[Function: handleEditorWidthChange]} handleSideContentHeightChange={[Function: handleSideContentHeightChange]} hasUnsavedChanges={false} mcqProps={{...}} sideContentHeight={[undefined]} sideContentProps={{...}} replProps={{...}} />
</div>"
`;
